### PR TITLE
#133: relation returns nil for differently zoned zdts

### PR DESCRIPTION
**Pull request by jgoday**
Define ITimeComparision === operator, to allow compare equality of ZonedDateTime.
Updated interval "Allen's Basic Relations" predicates to use this new operator (p/===).
Added new test case 'date-relation-test' to verify relations between differently zoned zdts.

### CODE REVIEW

**Comment by henryw374** [comment]
sorry, I missed this when it came in. I think it looks good, but will need a few changes since 0.5.0-RC1 because things have moved namespace. 

Hopefully @malcolmsparks can have a look as well.

**Comment by henryw374** [comment]
yes i think so. we can exclude `=` in the ns decl, and refer to it as clojure.core/= where it is used in tick.core

**Comment by jgoday** [comment]
> yes i think so. we can exclude `=` in the ns decl, and refer to it as clojure.core/= where it is used in tick.core

Ok, I have changed to '=', excluding clojure.core/= in the ns decl as you said (I'm not very used to it).

About the equality between instants and zdt,
should be enough to implement protocol = this way ? (comparing the instant value of both arguments)
```clojure

(extend-protocol p/ITimeComparison
...
  Instant
  (= [x y] (= (p/instant x) (p/instant y)))
...
  ZonedDateTime
  (= [x y] (= (p/instant x) (p/instant y)))
....

;; with the following test
(deftest zdt-and-instant-equality-test
  (testing "Instants and ZonedDateTimes should be equals if represents the same point in time"
    (is (p/=
            (t/instant (t/clock "2017-10-31T16:00:00Z"))
            (t/zoned-date-time "2017-10-31T16:00:00Z[UTC]")))))

```

File: src/tick/core.cljc [hunk 1]
```diff
@@ -514,51 +514,61 @@
   (<= [x y] (not (cljc.java-time.instant/is-after x y)))
   (> [x y] (cljc.java-time.instant/is-after x y))
   (>= [x y] (not (cljc.java-time.instant/is-before x y)))
+  (= [x y] (= x y))
```
Comment (henryw374) [inline src/tick/core.cljc:517]: = needs excluding at the top of this file

File: src/tick/core.cljc [hunk 2]
```diff
@@ -514,51 +514,61 @@
...
   LocalDateTime
   (< [x y] (cljc.java-time.local-date-time/is-before x y))
   (<= [x y] (not (cljc.java-time.local-date-time/is-after x y)))
   (> [x y] (cljc.java-time.local-date-time/is-after x y))
   (>= [x y] (not (cljc.java-time.local-date-time/is-before x y)))
+  (= [x y] (= x y))
```
Comment (henryw374) [inline src/tick/core.cljc:523]: so this becomes `(= [x y] (clojure.core/= x y))`

Comment (henryw374) [reply src/tick/core.cljc:523]: ah ignore this... i see it's only needed in the protocol ns

File: src/tick/core.cljc (hunk 1 above)
Comment (henryw374) [reply src/tick/core.cljc:517]: ah ignore this... i see it's only needed in the protocol ns

File: src/tick/core.cljc [hunk 3]
```diff
@@ -514,51 +514,61 @@
...
   (<= [x y] (not (cljc.java-time.local-time/is-after x y)))
   (> [x y] (cljc.java-time.local-time/is-after x y))
   (>= [x y] (not (cljc.java-time.local-time/is-before x y)))
+  (= [x y] (= x y))
   OffsetDateTime
   (< [x y] (cljc.java-time.offset-date-time/is-before x y))
   (<= [x y] (not (cljc.java-time.offset-date-time/is-after x y)))
   (> [x y] (cljc.java-time.offset-date-time/is-after x y))
   (>= [x y] (not (cljc.java-time.offset-date-time/is-before x y)))
+  (= [x y] (= x y))
```
Comment (henryw374) [inline src/tick/core.cljc:547]: offsetdatetime, zoneddatetime and instants should all be comparable

**Comment by henryw374** [comment]
> > yes i think so. we can exclude `=` in the ns decl, and refer to it as clojure.core/= where it is used in tick.core
> 
> Ok, I have changed to '=', excluding clojure.core/= in the ns decl as you said (I'm not very used to it).
👍 

> 
> About the equality between instants and zdt,
> should be enough to implement protocol = this way ? (comparing the instant value of both arguments)
> 
> ```clojure
> (extend-protocol p/ITimeComparison
> ...
>   Instant
>   (= [x y] (= (p/instant x) (p/instant y)))
> ...
>   ZonedDateTime
>   (= [x y] (= (p/instant x) (p/instant y)))
> ....

yes I think that looks good. also do the same for offsetdatetime

**Comment by henryw374** [comment]
also for tests can we have one where e.g. differently offset zdts are =. like 5am in NY is equal to 3am in LA or whatever

**Comment by jgoday** [comment]
> also for tests can we have one where e.g. differently offset zdts are =. like 5am in NY is equal to 3am in LA or whatever

Implemented equality using p/instant for Instant/Date/OffsetDateTime/ZonedDateTime and add some test cases for comparing dates/instants with different timezones.

**Comment by henryw374** [comment]
Cool. It occurred to me that the other relations, <= etc should also work for these 4 types, I can't remember if they do already or not

**Comment by jgoday** [comment]
> Cool. It occurred to me that the other relations, <= etc should also work for these 4 types, I can't remember if they do already or not

Currently 
```clojure
(p/<= (t/zoned-date-time "2017-10-31T16:00:00-04:00[America/New_York]")
         (t/instant (t/clock "2017-10-31T16:00:00Z")))
```
throws (in jvm)
```
Exception: java.lang.ClassCastException: class java.time.Instant cannot be cast to class java.time.chrono.ChronoZonedDateTime (java.time.Instant and java.time.chrono.ChronoZonedDateTime are in module java.base of loader 'bootstrap')
```
in node it produces an error
```
#object[TypeError TypeError: other.toEpochSecond is not a function]
```

**Comment by henryw374** [comment]
Thanks for checking

**Comment by henryw374** [comment]
Thanks for all the input. I tweaked your stuff a bit and committed in #3f283ff

**Pull request closed without merging**